From 7b6025108e66ddf0056568afb2f0383eca880bae Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:05:33 -0400
Subject: [virt] kvm/ppc: book3s pr - Fix proto-VSID calculations

Message-id: <1410545655-205645-105-git-send-email-dzickus@redhat.com>
Patchwork-id: 94127
O-Subject: [RHEL7 PATCH 104/626] KVM: PPC: Book3S PR: Fix proto-VSID calculations
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 8ed7b7e9d2a4e208d3c94af3ddbc628317b30409
Author: Paul Mackerras <paulus@samba.org>
Date:   Sat Jun 22 17:13:32 2013 +1000

    KVM: PPC: Book3S PR: Fix proto-VSID calculations

    This makes sure the calculation of the proto-VSIDs used by PR KVM
    is done with 64-bit arithmetic.  Since vcpu3s->context_id[] is int,
    when we do vcpu3s->context_id[0] << ESID_BITS the shift will be done
    with 32-bit instructions, possibly leading to significant bits
    getting lost, as the context id can be up to 524283 and ESID_BITS is
    18.  To fix this we cast the context id to u64 before shifting.

    Signed-off-by: Paul Mackerras <paulus@samba.org>
    Signed-off-by: Alexander Graf <agraf@suse.de>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kvm/book3s_64_mmu_host.c b/arch/powerpc/kvm/book3s_64_mmu_host.c
index 176d3fd..8757a25 100644
--- a/arch/powerpc/kvm/book3s_64_mmu_host.c
+++ b/arch/powerpc/kvm/book3s_64_mmu_host.c
@@ -325,9 +325,9 @@ int kvmppc_mmu_init(struct kvm_vcpu *vcpu)
   return -1;
  vcpu3s->context_id[0] = err;
 
- vcpu3s->proto_vsid_max = ((vcpu3s->context_id[0] + 1)
+ vcpu3s->proto_vsid_max = ((u64)(vcpu3s->context_id[0] + 1)
       << ESID_BITS) - 1;
- vcpu3s->proto_vsid_first = vcpu3s->context_id[0] << ESID_BITS;
+ vcpu3s->proto_vsid_first = (u64)vcpu3s->context_id[0] << ESID_BITS;
  vcpu3s->proto_vsid_next = vcpu3s->proto_vsid_first;
 
  kvmppc_mmu_hpte_init(vcpu);
-- 
1.7.1